### ISSUE TITLE
[$500] BUG: Mac/Firefox - Opening the protect PDF modal and pressing TAB doesn’t focus on enter the password reported by @dhairyasenjaliya 

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12058]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to any conversation
2. Send Protected PDF (if required)
3. Open Protected PDF
4. Press TAB key simultaneously try to navigate to enter the password with TAB
5. Notice it will focus only on Close Modal /Download PDF

## Expected Result:
On pressing TAB it should be able to focus on enter the password

## Actual Result:
On pressing TAB it does not focus on enter the password

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web - Mac Firefox only

**Version Number:** 1.2.18-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/197225996-c89a3cca-4019-4e1c-8008-e1013d7f2508.mov

https://user-images.githubusercontent.com/43996225/197226295-9146d9c4-8228-4951-8049-589f12a3484f.mp4

PDF used : 
[password-protected.pdf](https://github.com/Expensify/App/files/9872294/password-protected.pdf)

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @dhairyasenjaliya 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1666236736080999

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by mallenexpensify** [comment]
@zanyrenney  new BZ chore SO is [here](https://stackoverflow.com/c/expensify/questions/14418), we want to ensure BZ stays assigned to all issues so we can get to BugZero,

**Comment by zanyrenney** [comment]
Hey @kavimuru - Are you created the password protected PDF in Adobe? That's the only way I see how to recreate this. If you can attach an example of the PDF to this GH issue, i can test this properly.

**Comment by kavimuru** [comment]
@zanyrenney Added the pdf used for this test in the GH.

**Comment by kavimuru** [comment]
@zanyrenney Password is 1234 and attached another PDF

**Comment by zanyrenney** [comment]
I can't reproduce this - is it still happening for you @kavimuru 
![2022-10-27_16-00-21 (1)](https://user-images.githubusercontent.com/56830058/198325962-5d63ab74-16bb-4808-bf5a-7ae83cb8a14d.gif)

**Comment by zanyrenney** [comment]
I can't reproduce this on Mac/Chrome - is it only happening on FireFox?
 
![2022-10-27_16-00-21 (1)](https://user-images.githubusercontent.com/56830058/198326420-35dffd6f-2179-479c-97a8-9b88ca7b09ce.gif)

**Comment by zanyrenney** [comment]
Thanks. Asking in slack if we should be testing and fixing these super edge cases: https://expensify.slack.com/archives/C01SKUP7QR0/p1666963361304399

**Comment by parasharrajat** [comment]
No proposals yet.

**Comment by zanyrenney** [comment]
waiting on proposals.

**Comment by huzaifa-99** [comment]
## Proposal

### Root Cause
This is happening because of the `TextLink` component used in `PDFInfoMessage`. 
https://github.com/Expensify/App/blob/17219657fd7a044abdd7b08adfa4717856113e59/src/components/PDFView/PDFInfoMessage.js#L32-L34

1. It is not capturing focus on FireFox/Safari with `Tab`
2. It captures focus on chrome with `Tab` but when enter is pressed the focus is lost without doing the actual press/click and moving to password screen 

### Solution
Wrapping TextLink in `TouchableOpacity` fixes the issue
```js
<TouchableOpacity
    onPress={props.onShowForm}
>
    <TextLink onPress={props.onShowForm}>
        {` ${props.translate('attachmentView.pdfPasswordForm.linkText')} `}
    </TextLink>
</TouchableOpacity>
```

### Demo with proposed solution

https://user-images.githubusercontent.com/68777211/199773111-66233dba-d9d1-45a5-a96d-88be613fd088.mp4

### Suggestion
I believe the `TextLink` component should not be used here because it seems that it is used to render links and **enter the password** doesn't seem to be a link. 

Currently hovering over it with a mouse points to the current url address (I don't know if this is the expected behaviour). Notice the bottom left side and the current url address, in this demo

https://user-images.githubusercontent.com/68777211/199774077-e8c507fc-3613-4891-9922-096103fd3607.mp4

We can remove this (url address on bottom left) by using a normal `Text` component wrapped inside `TouchableOpacity` (also applying the link style on Text)
```js
<TouchableOpacity
    onPress={props.onShowForm}
>
    <Text style={[styles.link]}>
        {` ${props.translate('attachmentView.pdfPasswordForm.linkText')} `}
    </Text>
</TouchableOpacity>
``` 

`Pressable` would also work in place of `TouchableOpacity` if that is preferred.

---
Please let me know your thoughts @zanyrenney @parasharrajat

**Comment by huzaifa-99** [comment]
> Waiting on proposals, tomorrow we can up the price!

Oops!. Didn't saw this, I should have waited till tomorrow 😂

**Comment by parasharrajat** [comment]
The only reason that `TextLink` is used is that it does not create a block element in the middle of the text. @huzaifa-99 I liked your suggestion of using Pressable instead of this. But it will create the mentioned problem. I think the same goes for `TouchableOpacity`. 

But I agree using either of these will fix both this issue and https://github.com/Expensify/App/issues/11354

**Comment by parasharrajat** [comment]
> The only reason that TextLink is used is that it does not create a block element in the middle of the text. Block view will break the text wrapping and alignment.


@huzaifa-99 what do you think the best solution might be to fix this without causing this issue?

**Comment by michaelhaxhiu** [comment]
Hey @zanyrenney  just dropping a note as a reminder to [keep the pressure on](https://expensify.slack.com/archives/C01SKUP7QR0/p1667064948124109) to find a contributor and get this one closed out. If it helps incentivize @huzaifa-99 to finish the proposal and submit a PR sooner, I would even consider doubling the price now (even with an active proposal in review) 😉

**Comment by zanyrenney** [comment]
@huzaifa-99 and @parasharrajat price is doubled, hopefully we can get this one reviewed and submit the PR sooner 🙏🏼

**Comment by parasharrajat** [comment]
WAiting for proposals.

**Comment by roryabraham** [comment]
> Pressable would also work in place of TouchableOpacity if that is preferred.

Yes, `Pressable` is preferred.

https://github.com/Expensify/App/pull/12034 was just deployed to staging btw, so the `Enter` issue shouldn't exist anymore. Can we just add `focusable=true` to the `TextLink` component [here](https://github.com/Expensify/App/blob/d0ad7de988a88bcf4f2de61053968ba9c9faad0f/src/components/TextLink.js#L60-L68)?

**Comment by parasharrajat** [comment]
Yeah, that should do it.

**Comment by roryabraham** [comment]
Separately, why is this link even a thing? Why not just show the text input right away???

**Comment by parasharrajat** [comment]
BTW, I am able to focus on the link without any changes. But pressing `Enter` when the link is focused does not open the form, instead, submit the Attachment. This is happening due to the Enter key handler in the capture phase on Send button.

**Comment by roryabraham** [comment]
Thanks for that context, @parasharrajat.

> BTW, I am able to focus on the link without any changes

So you're not able to reproduce this issue, even on Firefox?

> But pressing Enter when the link is focused does not open the form, instead, submit the Attachment. This is happening due to the Enter key handler in the capture phase on Send button.

@parasharrajat Can you please bring this up in [this issue](https://github.com/Expensify/App/issues/11354) instead?

**Comment by parasharrajat** [comment]
Yeah, seems to be working fine. 

https://user-images.githubusercontent.com/24370807/200065129-68db6ebe-ce41-4cc7-b625-238bebdd5ad4.mp4

**Comment by zanyrenney** [comment]
^ waiting on retest.

**Comment by huzaifa-99** [comment]
@parasharrajat apology for the late reply.
> Block view will break the text wrapping and alignment on native.

I wanted to test how bad it looks (before updating the proposal) but got stuck on building for android/ios (still stuck), I mentioned it [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1667573885470069) in some detail. Please let me know if I am missing something.

**Comment by huzaifa-99** [comment]
> https://github.com/Expensify/App/pull/12034 was just deployed to staging btw, so the Enter issue shouldn't exist anymore. Can we just add focusable=true to the TextLink component [here](https://github.com/Expensify/App/blob/d0ad7de988a88bcf4f2de61053968ba9c9faad0f/src/components/TextLink.js#L60-L68)?

I think #12034 fixed the issue on Safari, but I am still able to reproduce it locally on FireFox (commit sha `0724337138a4d80f6603074cc9af5c186cb524ec`)

https://user-images.githubusercontent.com/68777211/200304642-c2700ae2-fc44-4e5d-969b-b4e70e6cbf41.mp4


Also adding `focusable=true` to `TextLink` works for me in Firefox, but only If `accessibilityRole` and `href` are not explicitly set (I commented them in my local test to make it work)

https://github.com/Expensify/App/blob/d0ad7de988a88bcf4f2de61053968ba9c9faad0f/src/components/TextLink.js#L60-L68

**Comment by parasharrajat** [comment]
I will give it a spin when possible. Meanwhile, you are free to propose if you are able to reproduce this issue.

**Comment by parasharrajat** [comment]
> I think https://github.com/Expensify/App/pull/12034 fixed the issue on Safari, but I am still able to reproduce it locally on FireFox (commit sha 0724337138a4d80f6603074cc9af5c186cb524ec)

Could you please take this on the [other](https://github.com/Expensify/App/issues/11354) issue?

**Comment by huzaifa-99** [comment]
> @huzaifa-99 I think you are referring to this particular issue https://github.com/Expensify/App/issues/12058 as reproducible locally right?

Yes. I am also able to reproduce this issue on staging `v1.2.24-0` with Firefox

**Comment by parasharrajat** [comment]
oh, ok. I thought we were discussing other issue. Yeah, both the different.

**Comment by zanyrenney** [comment]
Issue is here: https://www.upwork.com/jobs/~010f0af8b1df252e1d 

But @roryabraham can you advise on the above - should it be linked to another one of these issues?

**Comment by huzaifa-99** [comment]
> > The only reason that TextLink is used is that it does not create a block element in the middle of the text. Block view will break the text wrapping and alignment on native.
> 
> @huzaifa-99 what do you think the best solution might be to fix this without causing this issue?

### Updated Proposal (Original Proposal https://github.com/Expensify/App/issues/12058#issuecomment-1302344519)
We have to separate `Text` blocks and wrap it in a `View`

Basically, replacing this code block
https://github.com/Expensify/App/blob/cb93661530860f65e4aae0bc7ba7fe3c8db3e643/src/components/PDFView/PDFInfoMessage.js#L30-L36

with 
```js
<View style={{flexDirection: 'row'}}>
    <Text>
        {props.translate('attachmentView.pdfPasswordForm.beforeLinkText')}
    </Text>
    <Pressable onPress={props.onShowForm}>
        <Text style={[styles.link]}>
            {` ${props.translate('attachmentView.pdfPasswordForm.linkText')} `}
        </Text>
    </Pressable>
    <Text>
        {props.translate('attachmentView.pdfPasswordForm.afterLinkText')}
    </Text>
</View>
```
which fixes the issue on all platforms. 

Notice I have replaced `TextLink` with `Text` because of this reason mentioned in previous proposal https://github.com/Expensify/App/issues/12058#issuecomment-1302344519
> I believe the TextLink component should not be used here because it seems that it is used to render links and enter the password doesn't seem to be a link.

Both `Text` and `TextLink` work, please let me know what's preferred.

### IOS Demo

https://user-images.githubusercontent.com/68777211/200619490-137aaec3-2ab7-47f3-a551-2e14a4ba8a72.mp4

### Android Demo

https://user-images.githubusercontent.com/68777211/200619624-a5ee21b6-09c4-4675-a878-3e81fd60c68d.mp4

### Desktop Demo

https://user-images.githubusercontent.com/68777211/200619695-e0d9da1b-023f-473c-9167-28ca2f890e9c.mp4

### Chrome Demo

https://user-images.githubusercontent.com/68777211/200620960-3927dce6-0acb-435b-b89a-f9f6a3bef46c.mp4

### Firefox Demo

https://user-images.githubusercontent.com/68777211/200619831-502c7722-6f99-47fb-a6bb-beaf2474b66b.mp4

### Safari Demo

https://user-images.githubusercontent.com/68777211/200619806-455a992c-2b06-4bbc-aa12-4e7b0b407c44.mp4

---

To add I had faced #12480 while testing on iOS/Android which is totally unrelated to the current issue. I applied the fix for it to make iOS/Android demos. I saw @parasharrajat you were already there so just providing more context from my side.

Please let me know your thoughts.

**Comment by zanyrenney** [comment]
How is the review going @parasharrajat

**Comment by parasharrajat** [comment]
Checking now...

**Comment by parasharrajat** [comment]
Ok, this bug is occurring on the Firefox Mac but Firefox linux is working fine. :confused:.

**Comment by huzaifa-99** [comment]
which distro did you try?

**Comment by parasharrajat** [comment]
Arch Linux
---

Ok, I think we can go with @huzaifa-99 's [proposal](https://github.com/Expensify/App/issues/12058#issuecomment-1307500529).  As the change is applied to this particular text which does not wrap on any platform, it will work fine. 

I think there is no need to revert the https://github.com/Expensify/App/pull/12034. It will be useful for `TextLink Context`.

cc: @roryabraham 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by s77rt** [comment]
@huzaifa-99 
is it required to wrap in a `View`?

**Comment by huzaifa-99** [comment]
@s77rt yup, otherwise

![Screenshot 2022-11-11 at 11 57 48 PM](https://user-images.githubusercontent.com/68777211/201411740-1b3f1ec0-2a2a-4a9a-b71c-27c0bb9a2040.png)

Basically we have to render in a single row (wrapped in a View with flexDirection of row)

**Comment by s77rt** [comment]
I meant something like this
```diff
diff --git a/src/components/PDFView/PDFInfoMessage.js b/src/components/PDFView/PDFInfoMessage.js
index 5d583db40..898983f13 100644
--- a/src/components/PDFView/PDFInfoMessage.js
+++ b/src/components/PDFView/PDFInfoMessage.js
@@ -1,6 +1,6 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import {View} from 'react-native';
+import {View, Pressable} from 'react-native';
 import Text from '../Text';
 import TextLink from '../TextLink';
 import Icon from '../Icon';
@@ -29,9 +29,11 @@ const PDFInfoMessage = props => (
         <Text>{props.translate('attachmentView.pdfPasswordForm.infoText')}</Text>
         <Text>
             {props.translate('attachmentView.pdfPasswordForm.beforeLinkText')}
-            <TextLink onPress={props.onShowForm}>
-                {` ${props.translate('attachmentView.pdfPasswordForm.linkText')} `}
-            </TextLink>
+            <Pressable onPress={props.onShowForm}>
+                <Text style={[styles.link]}>
+                    {` ${props.translate('attachmentView.pdfPasswordForm.linkText')} `}
+                </Text>
+            </Pressable>
             {props.translate('attachmentView.pdfPasswordForm.afterLinkText')}
         </Text>
     </View>
```

**Comment by huzaifa-99** [comment]
@s77rt something similar already proposed [here](https://github.com/Expensify/App/issues/12058#issuecomment-1302344519) but this would not as suggested by @parasharrajat [here](https://github.com/Expensify/App/issues/12058#issuecomment-1302662198)

**Comment by zanyrenney** [comment]
Hi everyone - this issue is a duplicate of https://github.com/Expensify/App/issues/12010, please can you post your proposals on the original issue to avoid duplication and track this correctly. @huzaifa-99 @s77rt cc. @parasharrajat @roryabraham

**Comment by zanyrenney** [comment]
Closing this issue in favour of https://github.com/Expensify/App/issues/12010

**Comment by huzaifa-99** [comment]
@zanyrenney #12010 is related to the focus issues Download/Close buttons on Mac/Safari only. This issue is related to focus issues on a label "Enter the password" in Mac/FireFox only.

Should we still consider these as one?

**Comment by dhairyasenjaliya** [comment]
hello @zanyrenney both issues are separate 

#12058 this issue is about "enter the password" not focusing on the tab

<img width="1153" alt="Screenshot 2022-11-14 at 5 08 59 PM" src="https://user-images.githubusercontent.com/47522946/201650907-72b2d0eb-e782-4aa9-acdb-7e62e4add83c.png">


#12010 this issue is about protect the PDF modal and pressing TAB doesn’t focus on download/close
<img width="1139" alt="Screenshot 2022-11-14 at 5 10 07 PM" src="https://user-images.githubusercontent.com/47522946/201651086-39acc5d8-77d3-4920-81db-fbbbbe0632c8.png">

**Comment by s77rt** [comment]
@zanyrenney @huzaifa-99 @dhairyasenjaliya 
not the same issue.
but #12010 covers both #12058 and #12554

**Comment by huzaifa-99** [comment]
Bump @zanyrenney

**Comment by zanyrenney** [comment]
Sorry @huzaifa-99 but from this comment it looks like it is correct this is closed in favour of that issue that covers this off? https://github.com/Expensify/App/issues/12058#issuecomment-1313568732 @roryabraham please also let me know your thoughts here.

**Comment by dhairyasenjaliya** [comment]
@zanyrenney also bug reporting bonus remaining since it was valid at reporting time then later on it was decided not to focus on keyboard navigation issues

**Comment by s77rt** [comment]
The solution to this issue https://github.com/Expensify/App/issues/12010
is the solution of https://github.com/Expensify/App/issues/12058 and https://github.com/Expensify/App/issues/12554

I don't know how to handle such case but it's logical to close the new issues in favour of the old one but this didn't happen.
If it was about implementing a solution, I'd say @huzaifa-99 should be credited for solving a part of the issue (original issue), but I don't know how the reporting bonus work in such case so I can't help in that.

**Comment by JmillsExpensify** [comment]
@roryabraham @parasharrajat Is this issue still aimed at solving focus for TAB navigation? If so, I think we should close it. We've announced that these are new accessibility features and not bugs.

**Comment by roryabraham** [comment]
I agree, we should close this.

> also bug reporting bonus remaining since it was valid at reporting time then later on it was decided not to focus on keyboard navigation issues

Not sure if this is correct or not, but I think we need to sort this out when we close it. Normally reporting bonus is only paid for bugs and new features that are fixed/implemented.

**Comment by dhairyasenjaliya** [comment]
Alright, got it, thank you @JmillsExpensify @roryabraham I think we are good to close then since this issue is about TAB navigation

**Comment by parasharrajat** [comment]
No payment is due here.

**Comment by zanyrenney** [comment]
Taken to slack for discussion: https://expensify.slack.com/archives/C01GTK53T8Q/p1669809248604369

**Comment by zanyrenney** [comment]
hi @huzaifa-99 we decided that the proposal itself, which we approved is the only segment for payment, please can you apply for the job on upwork so we can pay you: https://www.upwork.com/jobs/~010f0af8b1df252e1d

**Comment by huzaifa-99** [comment]
@zanyrenney applied thanks. 

> we decided that the proposal itself, which we approved is the only segment for payment

Just to be sure, I should raise a PR, right?

**Comment by zanyrenney** [comment]
Hey @huzaifa-99 - no, you don't need to submit a PR. The reason you are due the payment is that you worked on the proposal already by the time we asked you not to do it, so, we're paying you for the proposal given the work was complete even though we aren't using it, for fairness and appreciation of your time. I hope that makes sense? Let me know if you have any questions.

**Comment by zanyrenney** [comment]
offer sent @huzaifa-99

**Comment by huzaifa-99** [comment]
> Hey @huzaifa-99 - no, you don't need to submit a PR. The reason you are due the payment is that you worked on the proposal already by the time we asked you not to do it, so, we're paying you for the proposal given the work was complete even though we aren't using it, for fairness and appreciation of your time. I hope that makes sense? Let me know if you have any questions.

@zanyrenney thanks for the consideration and followup :)

**Comment by zanyrenney** [comment]
Paid via upwork, closing this out. @huzaifa-99 Reopen if you have any problems or follow up questions.
